APE22.rst

```diff
@@ -0,0 +1,440 @@
+   - Add a SciPy-like section to the developer documentation explaining the
+     public vs internal rules.
+   - Fix links to always point to the public interface, not the internal
+     interface. (This will presumably require a similar implementation as used
+     in NumPy to change the ``__module__`` attribute of many objects.)
+   - Add a reminder to the maintainer checklist that to be public a symbol must be
+     in ``__all__`` and documented.
+   - Clearly state if a documented object is actually private.
+
+4. **Add a pre-commit bot**. This will ensure that the ``__all__`` attribute is
+   always present and up-to-date. It will also ensure that the documentation is
+   always up-to-date with the ``__all__`` attribute. This can be accomplished
+   with a pre-commit hook that runs a script that checks the ``__all__``
+   attribute and, if the module is public, searches in the ``docs/api``
+   directory for symbols in ``__all__``.
+
+5. **Add prefixes**: Add prefixes to the top-most private symbols. For modules
+   this makes all their contents private.
+
+In Astropy core each phase will be implemented on a per-sub-package basis with
+individual pull requests.
+
+
+Backward compatibility
+----------------------
+
+This APE breaks backward compatibility in two ways:
+
+1. Changes ``__all__`` in many modules. Many modules define an ``__all__`` but
+   include symbols that are intended to be private because they should be
+   imported from other modules (generally the top of the module). This will not
+   break code that directly imports the symbols (as Python does not use
+   ``__all__`` for this purpose), but it will break code that expects private
+   symbols and uses ``import *``.
+2. Adds prefixes to many symbols that are not public. This can be done in a
+   backward compatible way by adding a ``__getattr__`` method to the module that
+   raises a warning for any object that is not public. This will allow existing
+   code to continue to work, while encouraging people to fix their code to use
+   the public interface.
+
+
+Alternatives
+------------
+
+**We do nothing:**
+
+This is the status quo.  It is not a good option because it is not solve the
+issue.  The aforementioned problems of not i) knowing what is stable and
+supported, and what is not, remain.
+
+**We allow** ``__all__`` **to be optional:**
+
+This is not great.
+
+The only time this might be good is when a module has dynamic symbols from a
+`PEP 562 <https://peps.python.org/pep-0562/>`_ module-level ``__getattr__``
+method. However, if it truly dynamic then it cannot be statically analyzed, which
+is undesirable for other reasons. If it is not truly dynamic, then the "dynamic"
+symbols can be added to ``__all__`` and an ``__init__.pyi`` file used to
+communicate the public interface to static analyzers.
+
+**We make the documentation the authoritative definition of the API.**
+
+Good code is well documented in the code. Beyond code comments and docstrings, what is
```

Comment:
?
```suggestion
Good code is well documented. Beyond code comments and docstrings, what is
```

Reply:
No, I mean in the code itself. E.g. we build documentation from code docstrings because we want the docstrings to be available in the code itself. If that weren't important all our docstrings would just be in the docs.

Reply:
Oh, now I see what you meant !